tests/testthat/test-stats-panel.R

```diff
@@ -0,0 +1,53 @@
+test_that("get_active_users", {
+  log_data <- tibble::tibble(
+    time = strptime("2023-03-20 15:00:00", format = "%Y-%m-%d %H:%M:%S") + c(10, 20, 30, 3601),
+    username = c("a", "b", "a", "c"),
+    date = rep(Sys.Date(), 4)
+  )
+
+  result <- get_active_users(log_data)
+
+  # 2 different hours
+  expect_equal(NROW(result), 2)
+
+  # 2 userts on hour 3pm / 15h
+  expect_equal(
+    (result %>% dplyr::pull(.data$users))[1],
+    2
+  )
+})
+
+test_that("get_actions_per_day", {
+  log_data <- tibble::tibble(
+    action = c("login", "login", "logout", "something else", "other"),
+    date = rep(Sys.Date(), 5)
+  )
+
+  expect_equal(
+    get_actions_per_day(log_data) %>% dplyr::pull(.data$action),
+    2
+  )
+})
+
+# Change from tidyr::gather to tidyr::pivot_longer
+# for function: get_per_day_plot_data()
+test_that("gather pivot", {
+  dat <- tibble::tibble(
+    statistic = 1:10,
+    value = 11:20,
+    date = rep(Sys.Date(), 10)
+  )
+
+  gather_df <- dat %>%
+    tidyr::gather(key = "statistic", value = "value", -"date") %>%
+    dplyr::arrange(.data$date, .data$statistic, .data$value)
+
+  pivot_df <- dat %>%
+    tidyr::pivot_longer(c(-"date"), names_to = "statistic", values_to = "value") %>%
+    dplyr::arrange(.data$date, .data$statistic, .data$value)
+
+  expect_equal(
+    pivot_df,
+    gather_df
+  )
+})
```

Comment:
Thank you writing tests this early. We will do a lot of refactoring I think, and we will need to set up tests as fast as possible. I know these are only basic tests, but we can expand on these later.

Not quite sure we have to test it like this. `tidyr::gather()` has been superseded by `tidyr::pivot_longer()` anyway.

If we want to test, then we should check the output of `get_per_day_plot_data()` should return the expected output. I think this is related to the TDD tech talk about separating implementation from behavior.

Maybe something like this:

```
base <- data.frame(date = as.Date("2023-03-21"))

per_day <- tibble::tibble(
  date = as.Date("2023-03-21"),
  users = 3,
  sessions = 4,
  time = 0,
  actions = 4
)

per_day_plot_data <- data.frame(
  date = as.Date("2023-03-21"),
  statistic = c(
    "logged users (unique)", 
    "total opened sessions", 
    "avg session time (hours)", 
    "total clicks and inputs"
  ),
  value = c(3, 4, 0, 4),
  id = c(3, 1, 2, 1)
)
  
expect_equal(
  get_per_day_plot_data(base, per_day),
  per_day_plot_data
)
```

Reply:
Thanks for the suggestion, you're right that the test doesn't make sense. I'll remove it as it doesn't make sense outside the scope of local development.

I created it to make sure this change didn't have any impact:

https://github.com/Appsilon/shiny.telemetry/pull/20/files#diff-edfe75e84d29900eb4eebb0f04add9e13a381fe7e1e43d9b131ba8c5f95e213eL104-R128